### PR TITLE
Add a page for reserve transfer assets

### PR DESCRIPTION
**Pull request by 0x7CFE**
Initial draft for `assets_reserve_transfer` precompile

### CODE REVIEW

File: docs/xcm/building-with-xcm/xc-reserve-transfer.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
+---
+sidebar_position: 7
+---
+
+# Reserve Transfer Assets
+
+XCM allowus us to transfer assets from one chain to another in several ways. In this chapter we'll see how to use reserve transfers.
```
Comment (Dinonard) [inline docs/xcm/building-with-xcm/xc-reserve-transfer.md:7]: allowus --> allows us

File: docs/xcm/building-with-xcm/xc-reserve-transfer.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+First of all, you need to understand that we don't just magically teleport an asset to a remote chain, i.e. actual tokens do not leave the local chain, they just migrate to a special local account controlled by a remote entity. Basically we create a derivative that will represent local asset on a remote chain. This is done by registering an asset and mapping it's supply to a sovereign account. That way, the `assets` pallet will handle token minting and burning automatically, so the amount of minted derivatives will always match amount of actual tokens residing in corresponding sovereign account.
```
Comment (Dinonard) [inline docs/xcm/building-with-xcm/xc-reserve-transfer.md:9]: There are a couple of things that are incorrect here.

It's not us that creates a derivative, it's the destination chain. We don't have control over how they do this. There's no guarantee they're using `pallet-assets` at all or how they handle anything in that regard.

There is no actual mapping between an asset and a sovereign account. The concept of sovereign account is to have a privileged representation of a sibling parachain or relay chain locally in the form of an account.
The way it works is that when you use `reserve_transfer`, we can derive the destination parachain's sovereign account. But this isn't connected with an asset Id in any way.

Imo, this paragraph should either be rewritten or dropped.

File: docs/xcm/building-with-xcm/xc-reserve-transfer.md (hunk 1 above)
Comment (Dinonard) [inline docs/xcm/building-with-xcm/xc-reserve-transfer.md:7]: allowus --> allows us

File: docs/xcm/building-with-xcm/xc-reserve-transfer.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+# Transaction fees and asset sufficiency
+
+Every transaction must be paid. This is done to prevent transaction floods by creating economical pressure to sender. So, technically only those tokens that have some real value can be used to pay for transactions.
+
+Usually we pay for transactions using chain's native token. It's expected that its emission was controlled and will not cause any problems. But in some cases we may want to allow users pay for transactions using foreign assets only.
+
+To do that remote chain should be configured to allow XCM execution payment using that asset.
+
+If an asset is configured as sufficient then an account containing such an asset does not need to hold existential deposit in native tokens of that chain. So an account does not need to have any native tokens to receive such an asset.
```
Comment (Dinonard) [inline docs/xcm/building-with-xcm/xc-reserve-transfer.md:58]: We don't know whether the destination chain is using pallet-assets or even if they have a concept of `sufficient` assets.

File: docs/xcm/building-with-xcm/xc-reserve-transfer.md (hunk 2 above)
Comment (0x7CFE) [reply docs/xcm/building-with-xcm/xc-reserve-transfer.md:9]: > We don't have control over how they do this. There's no guarantee they're using pallet-assets at all or how they handle anything in that regard.

Good point. My original intention was to provide a context for the reader, i.e. what _could_ (but not necessarily _will_) happen. Thank you for the feedback.

Looks like currently I just mislead the reader.

**Comment by fiexer** [comment]
Any updates?

**Comment by 0x7CFE** [comment]
> Any updates?

Documentation was delayed a bit due to this PR being in progress https://github.com/AstarNetwork/astar-frame/pull/108

Now testing and updating the docs.

File: docs/xcm/building-with-xcm/xc-reserve-transfer.md [hunk 4]
```diff
@@ -47,12 +46,16 @@ This functionality is exposed to EVM smart contracts via precompiles. Interface
 
 **Note:** there is another version of `assets_reserve_transfer` precompile that accepts `address` instead of `bytes32` for `recipient_account_id`.
 
+# Transferring native assets
+
+Current API identifes assets being transferred by specifying an H160 style address (XC20). This prevents us from sending native token since there's no representation for it. However, there is a workaround for that which uses EVM `msg.value` API.
+
+Precompile implemenation checks `msg.value` and, if positive, treats it as another asset to be sent (`MultiLocation { parents: 0, interior: Here }`). In that case, native asset is added to the tail of `asset_id` and `asset_amount` lists and can be indexed by `fee_index` as any other asset in the list. Its value would be set equal to `msg.value`.
+
```
Comment (Dinonard) [inline docs/xcm/building-with-xcm/xc-reserve-transfer.md:54]: Would be good to have an example of this. E.g. both arrays for assets and amounts params are empty, but `.value` is greater than 0 which results in native currency transfer.

**Review by akru — APPROVED** [review]
That’s clean docs, good job!

**Pull request merged**